### PR TITLE
Added documentation regarding PodSecurityPolicies

### PR DESCRIPTION
**Pull request by christianhuening**
Added docs as requested in issue #1336 

[skip ci]

### CODE REVIEW

File: Documentation/cluster-crd.md [hunk 1]
```diff
@@ -295,3 +295,47 @@ spec:
           cpu: "2"
           memory: "4096Mi"
 ```
+
+## RBAC for PodSecurityPolicies
```
Comment (galexrt) [inline Documentation/cluster-crd.md]: As this is not a Cluster CRD option, I think this should only be put in the `kubernetes.md` so everything required is in the same file.

File: Documentation/kubernetes.md [hunk 2]
```diff
@@ -37,6 +37,71 @@ kubectl create -f rook-operator.yaml
...
+
+If you have activated the [PodSecurityPolicy Admission Controller](https://kubernetes.io/docs/admin/admission-controllers/#podsecuritypolicy) and thus are
+using [PodSecurityPolicies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/), you will require additional 
+(Cluster)RoleBindings for the different ServiceAccounts Rook uses to start the Rook Agent, CEPH Mons and OSD Pods.
+ 
+**Note**: You don not have to perform these steps if you do not have the PodSecurityPolicy Admission Controller activated!  
+
+##### PodSecurityPolicy 
+
+You will probably have a PodSecurityPolicy for elevated access in place, but you need one which allows hostNetwork usage and
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Could you put "PodSecurityPolicy" and "hostNetwork" in code blocks.
Also you should mention that allowing `hostNetwork` usage is only required when using `hostNetwork: true` in the Cluster CRD in the yaml below as a comment for `hostNetwork` with default `false` is good too.

File: Documentation/kubernetes.md [hunk 3]
```diff
@@ -37,6 +37,71 @@ kubectl create -f rook-operator.yaml
...
+    rule: RunAsAny
+  supplementalGroups:
+    rule: RunAsAny
+  volumes:
+  - '*'
+  allowedCapabilities:
+  - '*'
+  hostPID: true
+  hostIPC: true
+  hostNetwork: true
```
Comment (galexrt) [inline Documentation/kubernetes.md]: As written above and the `hostPorts` too.

File: Documentation/kubernetes.md [hunk 4]
```diff
@@ -37,6 +37,71 @@ kubectl create -f rook-operator.yaml
 kubectl -n rook-system get pod
 ```
 
+ 
+#### RBAC for PodSecurityPolicies
+
+If you have activated the [PodSecurityPolicy Admission Controller](https://kubernetes.io/docs/admin/admission-controllers/#podsecuritypolicy) and thus are
+using [PodSecurityPolicies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/), you will require additional 
+(Cluster)RoleBindings for the different ServiceAccounts Rook uses to start the Rook Agent, CEPH Mons and OSD Pods.
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Could you please put `ClusterRoleBindings` and `ServiceAccount`s in a code block.

Also it isn't only the Ceph Mons and OSD Pods, writing overall the "rook storage Pods" for example is better for the future features.

File: Documentation/kubernetes.md [hunk 5]
```diff
@@ -37,6 +37,71 @@ kubectl create -f rook-operator.yaml
...
+ 
+**Note**: You don not have to perform these steps if you do not have the PodSecurityPolicy Admission Controller activated!
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Please add code block around `PodSecurityPolicy`

File: Documentation/kubernetes.md [hunk 6]
```diff
@@ -37,6 +37,71 @@ kubectl create -f rook-operator.yaml
...
+  hostIPC: true
+  hostNetwork: true
+  hostPorts:
+  - min: 1
+    max: 65536
+```
+
+##### ClusterRoleBinding
+
+Next up you require a ClusterRoleBinding, which enables the Rook Agent ServiceAccount to run the rook-agent Pods on all nodes
```
Comment (galexrt) [inline Documentation/kubernetes.md]: As above add code blocks around the Kubernetes types.

File: Documentation/kubernetes.md [hunk 7]
```diff
@@ -37,6 +37,71 @@ kubectl create -f rook-operator.yaml
...
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: privileged-psp-user
+subjects:
+- kind: ServiceAccount
+  name: rook-agent
+  namespace: rook-system
+``` 
+
+You will also require two more RoleBindings for each Rook Cluster you deploy. More info can be found in the
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Here too.

**Comment by christianhuening** [comment]
I updated the branch with changed according to your comments. Hope I didn't miss anything.

File: Documentation/kubernetes.md [hunk 8]
```diff
@@ -36,6 +36,101 @@ kubectl create -f rook-operator.yaml
...
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: privileged-psp-user
+subjects:
+- kind: ServiceAccount
+  name: rook-ceph-osd
+  namespace: rook
+``` 
+
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Could you please remove that second empty line.

File: Documentation/kubernetes.md [hunk 9]
```diff
@@ -36,6 +36,101 @@ kubectl create -f rook-operator.yaml
...
 kubectl -n rook-system get pod
 ```
+## RBAC for PodSecurityPolicies
+
+If you have activated the [PodSecurityPolicy Admission Controller](https://kubernetes.io/docs/admin/admission-controllers/#podsecuritypolicy) and thus are
+using [PodSecurityPolicies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/), you will require additional `(Cluster)RoleBindings` 
+for the different `ServiceAccounts` Rook uses to start the Rook Storage Pods.
+ 
+**Note**: You don not have to perform these steps if you do not have the `PodSecurityPolicy` Admission Controller activated!  
+
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Could you please remove that second empty line.

File: Documentation/kubernetes.md [hunk 10]
```diff
@@ -36,6 +36,101 @@ kubectl create -f rook-operator.yaml
...
+    rule: RunAsAny
+  supplementalGroups:
+    rule: RunAsAny
+  volumes:
+  - '*'
+  allowedCapabilities:
+  - '*'
+  hostPID: true
+  hostIPC: true
+  hostNetwork: false
```
Comment (galexrt) [inline Documentation/kubernetes.md]: I haven't worked with PSPs yet, but aren't the `hostPorts` also needed with `hostNetwork: true`?

Comment (christianhuening) [reply Documentation/kubernetes.md]: yeah you're right. `hostNetwork` without `hostPorts`, is rather useless

File: Documentation/kubernetes.md [hunk 11]
```diff
@@ -36,6 +36,105 @@ kubectl create -f rook-operator.yaml
 # verify the rook-operator and rook-agents pods are in the `Running` state before proceeding
 kubectl -n rook-system get pod
 ```
+## RBAC for PodSecurityPolicies
+
+If you have activated the [PodSecurityPolicy Admission Controller](https://kubernetes.io/docs/admin/admission-controllers/#podsecuritypolicy) and thus are
+using [PodSecurityPolicies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/), you will require additional `(Cluster)RoleBindings` 
+for the different `ServiceAccounts` Rook uses to start the Rook Storage Pods.
+ 
+**Note**: You don not have to perform these steps if you do not have the `PodSecurityPolicy` Admission Controller activated!
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Typo `don not` -> `don't` or `do not`

File: Documentation/kubernetes.md [hunk 12]
```diff
@@ -36,6 +36,105 @@ kubectl create -f rook-operator.yaml
...
+
+If you have activated the [PodSecurityPolicy Admission Controller](https://kubernetes.io/docs/admin/admission-controllers/#podsecuritypolicy) and thus are
+using [PodSecurityPolicies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/), you will require additional `(Cluster)RoleBindings` 
+for the different `ServiceAccounts` Rook uses to start the Rook Storage Pods.
+ 
+**Note**: You don not have to perform these steps if you do not have the `PodSecurityPolicy` Admission Controller activated!  
+
+##### PodSecurityPolicy 
+
+You will probably have a `PodSecurityPolicy` for elevated access in place, but you need one privileged Pod execution. Here is an example:
```
Comment (galexrt) [inline Documentation/kubernetes.md]: "[...]but you need one privileged Pod execution" do you mean "you need one `PodSecurityPolicy` that allows privileged Pod execution"

File: Documentation/kubernetes.md [hunk 13]
```diff
@@ -36,6 +36,105 @@ kubectl create -f rook-operator.yaml
...
+  kind: ClusterRole
+  name: privileged-psp-user
+subjects:
+- kind: ServiceAccount
+  name: rook-agent
+  namespace: rook-system
+``` 
+
+You will also require two more `RoleBindings` for each Rook Cluster you deploy:
+Create these two `RoleBindings` in the Namespace you plan to deploy your Rook Cluster into (defaults to namespace "rook"):
```
Comment (galexrt) [inline Documentation/kubernetes.md]: "defaults to namespace"rook"" -> "default is "rook" namespace" sounds better.

File: Documentation/kubernetes.md [hunk 14]
```diff
@@ -36,6 +36,105 @@ kubectl create -f rook-operator.yaml
...
+
+```yaml
+# Allow the rook-agent serviceAccount to use the privileged PSP
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: rook-agent-psp
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
```
Comment (galexrt) [inline Documentation/kubernetes.md]: Where do you create the `ClusterRole` and `Role` named `privileged-psp-user` or does the user need to create it but how?

**Comment by christianhuening** [comment]
Updated again with changes. I totally missed that I didn't add the ClusterRole for the PSP, thanks for pointing it out!

File: Documentation/kubernetes.md [hunk 15]
```diff
@@ -36,6 +36,126 @@ kubectl create -f rook-operator.yaml
...
+  hostIPC: true
+  hostNetwork: false
+```
+
+**Hint**: Allowing `hostNetwork` usage is required when using `hostNetwork: true` in the Cluster `CustomResourceDefinition`!
+You will then also require to allow the usage of `hostPorts` in the `PodSecurityPolicy`, i.e:
+ ```yaml
+ hostPorts:
+   - min: 1
+     max: 65536
```
Comment (galexrt) [inline Documentation/kubernetes.md]: I checked the Ceph docs (see http://docs.ceph.com/docs/master/rados/configuration/network-config-ref/) and could you please change the `hostPorts` `min` to `6789` and `max` to `7300`.

**Comment by bassam** [comment]
We need to revisit some of these after #1314 is resolved.

**Pull request merged**